hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AssociateTokenRecipientsStep.java

```diff
@@ -0,0 +1,100 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.token.impl.handlers.transfer;
+
+import static com.hedera.hapi.node.base.ResponseCodeEnum.INVALID_ACCOUNT_ID;
+import static com.hedera.hapi.node.base.ResponseCodeEnum.TOKEN_NOT_ASSOCIATED_TO_ACCOUNT;
+import static com.hedera.node.app.service.token.impl.util.TokenHandlerHelper.getIfUsable;
+import static com.hedera.node.app.spi.workflows.HandleException.validateTrue;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.base.TokenID;
+import com.hedera.hapi.node.state.token.Token;
+import com.hedera.hapi.node.token.CryptoTransferTransactionBody;
+import com.hedera.node.app.service.token.impl.WritableAccountStore;
+import com.hedera.node.app.service.token.impl.WritableTokenRelationStore;
+import com.hedera.node.app.service.token.impl.WritableTokenStore;
+import com.hedera.node.app.service.token.impl.handlers.BaseTokenHandler;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Associates the token with the sender and receiver accounts if they are not already associated.
+ * They are auto-associated only if there are open auto-associations available on the account.
+ */
+public class AssociateTokenRecipientsStep extends BaseTokenHandler implements TransferStep {
+    private final CryptoTransferTransactionBody op;
+
+    public AssociateTokenRecipientsStep(@NonNull final CryptoTransferTransactionBody op) {
+        this.op = requireNonNull(op);
+    }
+
+    @Override
+    public void doIn(@NonNull final TransferContext transferContext) {
+        requireNonNull(transferContext);
+        final var handleContext = transferContext.getHandleContext();
+        final var tokenStore = handleContext.writableStore(WritableTokenStore.class);
+        final var tokenRelStore = handleContext.writableStore(WritableTokenRelationStore.class);
+        final var accountStore = handleContext.writableStore(WritableAccountStore.class);
+
+        for (final var xfers : op.tokenTransfersOrElse(emptyList())) {
+            final var tokenId = xfers.tokenOrThrow();
+            final var token = getIfUsable(tokenId, tokenStore);
+
+            for (final var aa : xfers.transfersOrElse(emptyList())) {
+                final var accountId = aa.accountID();
+                validateAndAutoAssociate(accountId, tokenId, token, accountStore, tokenRelStore, handleContext);
+            }
+
+            for (final var aa : xfers.nftTransfersOrElse(emptyList())) {
+                final var receiverId = aa.receiverAccountID();
+                final var senderId = aa.senderAccountID();
+                validateAndAutoAssociate(senderId, tokenId, token, accountStore, tokenRelStore, handleContext);
```

Comment:
It doesn't make sense to auto-associate the sender, does it? (Since they must already own the NFT here.)

Reply:
yeah, had same question. But `mono-service ` does it for all balance changes [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/tokens/TokenStore.java#L98). Should I remove this ?

Reply:
Removed.